[WebGPU] Pipelines should check if createFunction failed

tadeuzagallo opened the pull request:
#### 61646fa2b75cc97a71988d7fdad17f3f9109b146
<pre>
[WebGPU] Pipelines should check if createFunction failed
<a href="https://bugs.webkit.org/show_bug.cgi?id=256964">https://bugs.webkit.org/show_bug.cgi?id=256964</a>
rdar://109515339

Reviewed by Mike Wyrzykowski.

If `createFunction` returns nil, assiging it to the Metal pipeline descriptor will
result in a crash. Instead, we should return an invalid pipeline so we can report
an error instead of crashing.

* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):

Canonical link: <a href="https://commits.webkit.org/264233@main">https://commits.webkit.org/264233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9772b4fe8c1efb3969884f1a4b026804e63afb6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7263 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7442 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8635 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7254 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8525 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7195 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10163 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7147 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6477 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8738 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5183 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14160 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6847 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9342 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5698 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6324 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1679 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10519 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6707 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/ComputePipeline.mm:
```diff
@@ -83,6 +83,8 @@ static MTLSize metalSize(auto workgroupSize)
 
     auto function = createFunction(library, entryPointInformation, descriptor.compute.constantCount, descriptor.compute.constants, label);
+    if (!function)
+        return ComputePipeline::createInvalid(*this);
```
I think we are supposed to generate an internal error here too, but I can add that in a different PR if you prefer.

Would require this patch to Device.h / .mm and a call to `generateAnInternalError("...")`

```
diff --git a/Source/WebGPU/WebGPU/Device.h b/Source/WebGPU/WebGPU/Device.h
index 50de904ff871..a3a71c7f3df2 100644
--- a/Source/WebGPU/WebGPU/Device.h
+++ b/Source/WebGPU/WebGPU/Device.h
@@ -111,6 +111,7 @@ public:
 
     void generateAValidationError(String&& message);
     void generateAnOutOfMemoryError(String&& message);
+    void generateAnInternalError(String&& message);
 
     Instance& instance() const { return m_adapter->instance(); }
     bool hasUnifiedMemory() const { return m_device.hasUnifiedMemory; }
diff --git a/Source/WebGPU/WebGPU/Device.mm b/Source/WebGPU/WebGPU/Device.mm
index 02742ef83621..201569ec603f 100644
--- a/Source/WebGPU/WebGPU/Device.mm
+++ b/Source/WebGPU/WebGPU/Device.mm
@@ -287,6 +287,25 @@ void Device::generateAnOutOfMemoryError(String&& message)
     }
 }
 
+void Device::generateAnInternalError(String&& message)
+{
+    // https://gpuweb.github.io/gpuweb/#abstract-opdef-generate-an-internal-error
+
+    auto* scope = currentErrorScope(WGPUErrorFilter_Internal);
+
+    if (scope) {
+        if (!scope->error)
+            scope->error = Error { WGPUErrorType_Internal, WTFMove(message) };
+        return;
+    }
+
+    if (m_uncapturedErrorCallback) {
+        instance().scheduleWork([protectedThis = Ref { *this }, message = WTFMove(message)]() mutable {
+            protectedThis->m_uncapturedErrorCallback(WGPUErrorType_Internal, WTFMove(message));
+        });
+    }
+}
+
```

tadeuzagallo replied:
I'll let you add that in a separate PR, if that's ok. I'm not sure what is the right error to report here.

tadeuzagallo committed:
[WebGPU] Pipelines should check if createFunction failed
https://bugs.webkit.org/show_bug.cgi?id=256964
rdar://109515339

Reviewed by Mike Wyrzykowski.

If `createFunction` returns nil, assiging it to the Metal pipeline descriptor will
result in a crash. Instead, we should return an invalid pipeline so we can report
an error instead of crashing.

* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):

Canonical link: https://commits.webkit.org/264233@main
